File: frontend/models/ManufacturerModel.ts
/**
 * @license EUPL-1.2
 * SPDX-FileCopyrightText: 2024
 * - Nils Brinckmann <nils.brinckmann@gfz-potsdam.de>
 * - Marc Hanisch <marc.hanisch@gfz-potsdam.de>
 * - Helmholtz Centre Potsdam - GFZ German Research Centre for Geosciences (GFZ, https://www.gfz-potsdam.de)
 *
 * SPDX-License-Identifier: EUPL-1.2
 */

import { ExportControl } from '@/models/ExportControl'

export interface IManufacturerModel {
  id: string
  manufacturerName: string
  model: string
  externalSystemName: string
  externalSystemUrl: string

  exportControl: ExportControl | null
}

export class ManufacturerModel implements IManufacturerModel {
  private _id: string = ''
  private _manufacturerName: string = ''
  private _model: string = ''
  private _externalSystemName: string = ''
  private _externalSystemUrl: string = ''

  private _exportControl: ExportControl | null = null

  get id (): string {
    return this._id
  }

  set id (newId: string) {
    this._id = newId
  }

  get manufacturerName (): string {
    return this._manufacturerName
  }

  set manufacturerName (newManufacturerName: string) {
    this._manufacturerName = newManufacturerName
  }

  get model (): string {
    return this._model
  }

  set model (newModel: string) {
    this._model = newModel
  }

  get externalSystemName (): string {
    return this._externalSystemName
  }

  set externalSystemName (newExternalSystemName: string) {
    this._externalSystemName = newExternalSystemName
  }

  get externalSystemUrl (): string {
    return this._externalSystemUrl
  }

  set externalSystemUrl (newExternalSystemUrl: string) {
    this._externalSystemUrl = newExternalSystemUrl
  }

  get exportControl (): ExportControl | null {
    return this._exportControl
  }

  set exportControl (newExportControl: ExportControl | null) {
    this._exportControl = newExportControl
  }

  static createFromObject (someObject: IManufacturerModel): ManufacturerModel {
    const newObject = new ManufacturerModel()

    newObject.id = someObject.id
    newObject.manufacturerName = someObject.manufacturerName
    newObject.model = someObject.model
    newObject.externalSystemName = someObject.externalSystemName
    newObject.externalSystemUrl = someObject.externalSystemUrl

    newObject.exportControl = someObject.exportControl === null ? null : ExportControl.createFromObject(someObject.exportControl)

    return newObject
  }
}

File: backend/app/project/api/models/manufacturer_model.py
# SPDX-FileCopyrightText: 2024
# - Nils Brinckmann <nils.brinckmann@gfz-potsdam.de>
# - Helmholtz Centre Potsdam - GFZ German Research Centre for Geosciences (GFZ, https://www.gfz-potsdam.de)
#
# SPDX-License-Identifier: HEESIL-1.0

"""Model for manufacturer models."""

from ..es_utils import ElasticSearchIndexTypes, settings_with_ngrams
from .base_model import db
from .mixin import SearchableMixin


class ManufacturerModel(db.Model, SearchableMixin):
    """Combination of manufacturer name and model."""

    id = db.Column(db.Integer, primary_key=True, autoincrement=True)
    manufacturer_name = db.Column(db.String(256), nullable=False)
    model = db.Column(db.String(256), nullable=False)
    # The external system fields are there to point to different
    # sensor system that want to use the export control information
    # check from the SMS (GIPP for example).
    # The url is there so that we can link to that application
    # (as we don't list their devices within the SMS).
    external_system_name = db.Column(db.String(256), nullable=True)
    external_system_url = db.Column(db.String(256), nullable=True)

    def to_search_entry(self):
        """Convert the model to a dict to store in the full text search."""
        export_control = {}
        if self.export_control:
            export_control[
                "export_control_classification_number"
            ] = self.export_control.export_control_classification_number
            export_control[
                "customs_tariff_number"
            ] = self.export_control.customs_tariff_number
            export_control["dual_use"] = self.export_control.dual_use

        return {
            "manufacturer_name": self.manufacturer_name,
            "model": self.model,
            "external_system_name": self.external_system_name,
            "export_control": export_control,
        }

    @staticmethod
    def get_search_index_properties():
        """Get the properties for the index."""
        type_keyword_and_full_searchable = (
            ElasticSearchIndexTypes.keyword_and_full_searchable(
                analyzer="ngram_analyzer"
            )
        )
        return {
            "manufacturer_name": type_keyword_and_full_searchable,
            "model": type_keyword_and_full_searchable,
            "external_system_name": type_keyword_and_full_searchable,
            "export_control": {
                "type": "nested",
                "properties": {
                    "export_control_classification_number": type_keyword_and_full_searchable,
                    "customs_tariff_number": type_keyword_and_full_searchable,
                    "dual_use": {"type": "boolean"},
                },
            },
        }

    @classmethod
    def get_search_index_definition(cls):
        """Return the index configuration."""
        return {
            "aliases": {},
            "mappings": {"properties": cls.get_search_index_properties()},
            "settings": settings_with_ngrams(
                analyzer_name="ngram_analyzer",
                filter_name="ngram_filter",
                min_ngram=1,
                max_ngram=10,
                max_ngram_diff=10,
            ),
        }
